cpp/src/neighbors/detail/hnsw.hpp

```diff
@@ -206,6 +206,213 @@ std::enable_if_t<hierarchy == HnswHierarchy::CPU, std::unique_ptr<index<T>>> fro
+                host_dataset_view.data_handle() + (pt_id + 1) * host_dataset_view.extent(1),
+                host_query_set.data_handle() + (i - start_idx) * host_dataset_view.extent(1));
+    }
+    raft::copy(query_set.data_handle(),
+               host_query_set.data_handle(),
+               query_set.size(),
+               raft::resource::get_cuda_stream(res));
+
+    // initialize brute force index
+    auto bf_index = cuvs::neighbors::brute_force::build(
+      res, cuvs::neighbors::brute_force::index_params{}, query_set_view);
+
+    // search for nearest neighbors
+    auto neighbors = raft::make_device_matrix<int64_t, int64_t>(res, num_pts, neighbor_size);
+    auto distances = raft::make_device_matrix<float, int64_t>(res, num_pts, neighbor_size);
+    cuvs::neighbors::brute_force::search(res,
+                                         cuvs::neighbors::brute_force::search_params{},
+                                         bf_index,
+                                         query_set_view,
+                                         neighbors.view(),
+                                         distances.view());
+
+    auto host_neighbors_full =
+      raft::make_host_matrix<int64_t, int64_t>(neighbors.extent(0), neighbors.extent(1));
+    raft::copy(host_neighbors_full.data_handle(),
+               neighbors.data_handle(),
+               neighbors.size(),
+               raft::resource::get_cuda_stream(res));
+    raft::resource::sync_stream(res);
+
+    // Need to slice the first column of the neighbors matrix because it's the query point itself
+    auto host_neighbors =
+      raft::make_host_matrix<int64_t, int64_t>(neighbors.extent(0), neighbors.extent(1) - 1);
+#pragma omp parallel for
+    for (int64_t i = 0; i < host_neighbors.extent(0); i++) {
+      std::copy(host_neighbors_full.data_handle() + i * host_neighbors_full.extent(1) + 1,
+                host_neighbors_full.data_handle() + (i + 1) * host_neighbors_full.extent(1),
+                host_neighbors.data_handle() + i * host_neighbors.extent(1));
+    }
+
+    // add points to the HNSW index upper layers
+#pragma omp parallel for
+    for (auto i = start_idx; i < end_idx; i++) {
+      auto pt_id       = order[i];
+      auto internal_id = hnsw_internal_ids[pt_id];
+      auto ll_cur      = appr_algo->get_linklist(internal_id, pt_level);
+      appr_algo->setListCount(ll_cur, host_neighbors.extent(1));
+      auto* data     = (uint32_t*)(ll_cur + 1);
+      auto neighbors = host_neighbors.data_handle() + (i - start_idx) * host_neighbors.extent(1);
+      for (auto j = 0; j < host_neighbors.extent(1); j++) {
+        auto neighbor_id          = order[neighbors[j] + start_idx];
+        auto neighbor_internal_id = hnsw_internal_ids[neighbor_id];
+        data[j]                   = neighbor_internal_id;
+      }
+    }
+  }
+
+  // move cagra graph to host
+  auto graph = cagra_index.graph();
```

Comment:
Would be good to check whether the graph is accessible on the host, and only copy if it is not.